### ISSUE TITLE
Support CAR uploads 🚘✨⤴️

### ISSUE THREAD

**Issue by olizilla** [ipfs-shipyard/nft.storage#172]
[Content Addressed Archives (CAR)](https://github.com/ipld/specs/blob/master/block-layer/content-addressable-archives.md) package up blocks with their CIDs. Users could pre-compute the root CID for their assets by creating the CAR file locally and uploading that. nft.storage would store the exact set of blocks provided in the CAR, and the single root CID listed in the CAR header would be used as the asset CID. 

This removes the need to trust nft.storage. The client chunks and hashes the file to derive the CID, nft.storage just stores it as provided. A local IPFS node can be used to create a CAR via `ipfs dag export <root CID>` or tools like [`ipfs-car`](https://github.com/vasco-santos/ipfs-car) or the web based https://car.on.fleek.co

It also allows users to begin minting contracts with the CID they created and upload the CAR file to nft.storage in parallel. They would be responsible for preserving a copy of the CAR until the upload to nft.storage is confirmed, so it can be retried in the event of an error, but the guarantee of content addressing means that they can safely use the CID they created in a contract, and that same CID will be available over IPFS once the CAR uploads and import completes.

## TODO

- [x] add car upload test to [packages/api/test/nfts-upload.spec.js](https://github.com/ipfs-shipyard/nft.storage/blob/main/packages/api/test/nfts-upload.spec.js)
- [x] add api route `/uploadCar` or update `/upload` to check for a `Content-Type: application/car` header (or other flag, TBC)
	- [ ] verify the user provided CAR has a single root in the header.
	- [ ] (TBC) verify the CAR is complete? I assume cluster will try and find any missing blocks, but we probably don't want that behaviour, or I could upload a CAR with a single block for the root of wikipedia, and have cluster pull in a huge DAG.
	- [x] ensure the car blob passed to cluster.add has the content-type `application/car` so it is added to ipfs-cluster as a CAR rather than a regular file.
	- [ ] verify the [cid returned from cluster](https://github.com/ipfs-shipyard/nft.storage/blob/ff032f1928b2493d3e040f971732f81879489137/packages/api/src/routes/nfts-upload.js#L52) matches the root cid in the user provided CAR 
	- [x] pin the car file root to cluster if not already pinned. (TBC Does cluster do this automatically with a car file?)
	- [x] check if the user has previously uploaded this CID before
	- [x] if not, persist it via the nft model and add the CID to the pinataQueue.
	- [x] figure out how to update the openapi spec for the api as necessary [openapi schema](https://github.com/ipfs-shipyard/nft.storage/blob/main/packages/website/public/schema.yml)
	- [ ] regenerate the openapi clients if needed.
- [x] update js client `storeBlob` method to verify car file, or add a `storeCar` method
	- [ ] add client test for car files
	- [x] [example added](https://github.com/ipfs-shipyard/nft.storage/tree/main/packages/client/examples) to the JS client
- [ ] update docs as needed
- [x] update website to explain how to create an upload a CAR
- [x] nice to have... add a checkbox to file upload to say "this is a CAR", otherwise you end up with a CID for the CAR file itself.

**Comment by olizilla** [comment]
OpenAPI schema is defined here https://github.com/ipfs-shipyard/nft.storage/blob/main/packages/website/public/schema.yml

**Comment by alanshaw** [comment]
🚀 ✨

I'm just wondering how much of the CAR stuff we can offload onto cluster.

* Cluster already requires single root.
* I have the same assumption about Cluster automatically finding missing blocks from the CAR so yes we probably need to ensure it is complete.
* It's uploaded via the Cluster `/add` endpoint so I assume the root is pinned automatically (just like IPFS add). We're actually relying on this assumption already 😅.
* I don't know if Cluster already does CAR verification. I assume yes??

@hsanjuan do you have quick answers?

Once it's been sent to Cluster the code for checking existing, adding to NFTS/PINS/PINATA_QUEUE is already there and I don't think needs changes?

The OpenAPI schema is hand crafted. I think last time I used an online validator and then manually checked the generated docs looked sane. It would be cool to have an CI step to validate the schema.

[`client-gen.sh`](https://github.com/ipfs-shipyard/nft.storage/blob/main/client-gen.sh) is responsible for generating the clients in various languages.

Does the JS client need to do CAR verification? Can/should that be a feature `ipfs-car` provides (or a separate library)?

Please also can we have an [example added](https://github.com/ipfs-shipyard/nft.storage/tree/main/packages/client/examples) to the JS client?

**Comment by olizilla** [comment]
> Does the JS client need to do CAR verification? Can/should that be a feature ipfs-car provides (or a separate library)?

We can skip CAR verification in the client, and let the server tell is if something is wrong. We'd need to complete the refactor of the unixfs exporter to have a resonable bundle size for an http client, so yes, it's worth punting on client-side validation for this iteration. `ipfs-car` will produce complete CARs with a single root. 

We could add a verify option there to check that a provided CAR meets those expectations, but as they are not a requirement to be a valid CAR it might be a little awkward to explain. I think we can recommend folks use IPFS or `ipfs-car` to create their CARs, and let the server error if a given CAR is not up to the nft.storage standard.

**Comment by olizilla** [comment]
oh mime types https://github.com/ipld/specs/issues/368

**Comment by hsanjuan** [comment]
> I'm just wondering how much of the CAR stuff we can offload onto cluster.
> @hsanjuan do you have quick answers?

yes.

> * Cluster already requires single root.

The car header must have a single root in it. This is the CID that is pinned and is what cluster sends back.

> * I have the same assumption about Cluster automatically finding missing blocks from the CAR so yes we probably need to ensure it is complete.

Cluster does not inspect the CAR, just block-puts whatever is in it. If blocks are missing, or there are more blocks than needed, it will just block/put anything and then tell ipfs to pin the root. So yes, ipfs (not cluster), will try to find missing blocks.

> * It's uploaded via the Cluster `/add` endpoint so I assume the root is pinned automatically (just like IPFS add). We're actually relying on this assumption already sweat_smile.

Yes, cluster always pins.

> * I don't know if Cluster already does CAR verification. I assume yes??

To complete my question above, the ipld's CarReader code, verifies that the block data and the CID it travels with in the CAR file match. Other than that no verification is performed.

I think the CAR spec does not specify order of blocks. It would be nice if it did, because then perhaps you would be able to do it without reading the whole thing on memory first.

> 
>
